packages/stream_core/lib/src/utils/shared_emitter.dart

```diff
@@ -2,76 +2,199 @@ import 'dart:async';
 
 import 'package:rxdart/rxdart.dart';
 
-abstract class SharedEmitter<T> {
-  Future<E> waitFor<E extends T>({
-    required Duration timeLimit,
-  });
-
-  StreamSubscription<E> on<E extends T>(void Function(E event) onEvent);
-
+/// A read-only emitter that allows listening for events of type [T].
+///
+/// Listeners can subscribe to receive events, wait for specific event types,
+/// and register handlers for certain event types. The emitter supports
+/// type filtering, allowing listeners to only receive events of a specific
+/// subtype of [T].
+///
+/// See also:
+/// - [MutableSharedEmitter] for the mutable interface that allows emitting events.
+abstract interface class SharedEmitter<T> {
+  /// The stream of events emitted by this emitter.
+  ///
+  /// Returns a [Stream] that emits events of type [T].
+  Stream<T> get stream;
+
+  /// Waits for an event of type [E] to be emitted within the specified [timeLimit].
+  ///
+  /// If such an event is emitted, it is returned. If the time limit
+  /// is exceeded without receiving the event, a [TimeoutException] is thrown.
+  ///
+  /// Returns a [Future] that completes with the first event of type [E].
+  Future<E> waitFor<E extends T>({Duration? timeLimit});
+
+  /// Registers a handler [onEvent] that will be invoked whenever an event of type [E] is emitted.
+  ///
+  /// Returns a [StreamSubscription] that can be used to manage the subscription.
+  StreamSubscription<E> on<E extends T>(
+    void Function(E event) onEvent,
+  );
+
+  /// Returns the first element that satisfies the given [test].
+  ///
+  /// If no such element is found and [orElse] is provided, calls [orElse] and returns its result.
+  /// If no element is found and [orElse] is not provided, throws a [StateError].
+  ///
+  /// Returns a [Future] that completes with the first matching element.
   Future<T> firstWhere(
     bool Function(T element) test, {
-    required Duration timeLimit,
```

Comment:
I think we do use this time limit couple of times in video, quite useful to keep.